### DISCUSSION TITLE
How is testRuntime p2Installed of tycho surefire supposed to be used?

### DISCUSSION THREAD

**Discussion by phermsdorf** [eclipse-tycho/tycho#2357] [Q&A] [answered]
Hi,

what i want to achieve: 
* have a build which produces artifacts of a product / rcp application
* have another build (on another machine) which runs swtbot tests against this previously build product (to prevent that the whole product needs to be build again)

what i've tried:
run my build with `mvn clean verify` without test (or -DskipTests=true) to create the product artifacts. works
run the build again with `mvn test` (with the swtbot test bundles in the reactor) and a profile which configures:

```
<plugin>
    <groupId>org.eclipse.tycho</groupId>
    <artifactId>tycho-surefire-plugin</artifactId>
    <configuration>
        <testRuntime>p2Installed</testRuntime>
        <work>/pathto/product/</work>
    </configuration>
</plugin>
```
But as soon as i run the build with `mvn test` (even without the additional tycho-surefire configuration) the build starts to fail with missing dependencies to artifacts (which actually are in the reactor) but maybe because they where not compiled.
The error and info messages clearly indicate that tycho tries to download them from maven central:

```
[INFO] Building com.godyo.p5.database 5.5.2321-SNAPSHOT                [11/203]
[INFO] ---------------------------[ eclipse-plugin ]---------------------------
[INFO] Downloading from : https://repo.maven.apache.org/maven2/com/godyo/p4/com.godyo.p5.externlib.oracle/5.2.0/com.godyo.p5.externlib.oracle-5.2.0.jar

[ERROR] Failed to execute goal on project com.godyo.p5.database: Could not resolve dependencies for project com.godyo.p4:com.godyo.p5.database:eclipse-plugin:5.5.2321-SNAPSHOT: Could not find artifact com.godyo.p4:com.godyo.p5.externlib.oracle:jar:5.2.0 in central (https://repo.maven.apache.org/maven2) -> [Help 1]
```
com.godyo.p5.externlib.oracle is listed as module in the "Reactor Build Order"

So i don't come to the point where test execution would take place ...

Is this basically the correct setup and usage for my usecase?

Thanks for any hints on this!

Bye Peter

**Comment by laeubi** [comment]
@phermsdorf can you provide a small reproducer product? from the snippet it is not completely clear how do you try to use Tycho, e.g. in the log output I see `eclipse-plugin` is used but usually for such setup one uses `eclipse-test-plugin` or require some additional config/setup so maybe it is a configuration problem. Beside that, `p2Installed` does not mean you use an existing product and the work dir is where the workspace is located, so it seems a bit mixed up.

If you want to run the product, then you could use the eclipse-run plugin, but it seems you actually want a mixture of unit-test and running a product...

**Reply by phermsdorf** [reply]
I feel that I've a misunderstanding of what testRuntime p2Installed is used for.

When looking and running the integration test project `tycho/tycho-its/projects/surefire.p2InstalledRuntime` like the P2InstalledTestRuntimeTest does:

`mvn clean integration-test -PprovisionProduct -Dp2.repo.url=http://download.eclipse.org/releases/2023-03/ -Dtycho-version=3.0.4`
builds and creates the product.
The second run with
`mvn clean integration-test -PuseProvisionedProduct -DproductClassifier=linux.gtk.x86_64 -Dp2.repo.url=http://download.eclipse.org/releases/2023-03/ -Dtycho-version=3.0.4`

it also seems to build and package the whole product. I see "Building jar", "Building zip", "Compiling source file" etc.

So what is then the point to use `<testRuntime>p2Installed</testRuntime>` in the profile `useProvisionedProduct`?

I'm looking for a way to "build the test fragments and run the test against a p2 provisioned product" without building and packaging the whole product again. 
Is that possible?

Thanks for your comments

**Reply by phermsdorf** [reply]
> I'm looking for a way to "build the test fragments and run the test against a p2 provisioned product" without building and packaging the whole product again.

is that a valid request anyway? 
Wouldn't it mean that the compilation of the test fragments would need to happen against the already provisioned/provided product at a given directory path and additionally install everything else what's needed to execute the compiled tests (the tests themselfes, junit, swtbot etc.) from the target platform?

Thanks for any hints on this.

Bye Peter

**Reply by mickaelistria** [reply, marked as answer]
`p2Installed` will make tycho-surefire-plugin provision a full RCP application/product to run the tests, instead of using a minimal one from the target-platform as it does usually (which is still better in most cases though). The use-case is more for RCP product producers who do not care about other installations that their own product.
However, it's purely for tests, compilation and other are not affected. The right way to control what is used for dependency resolution and compilation would be to define the target platform.

**Comment by phermsdorf** [comment]
Thanks @mickaelistria for clarifying this.
So using kind of an "intermediate" target which includes the build product may be a way to acomplish what i want.

Thank you!
